### PR TITLE
[Mailer] Fix rendered templates for notifications

### LINKED ISSUE #48472: Update to 6.2 breaks email sending.
### Symfony version(s) affected

6.2

### Description

After updating to 6.2 I got this message:

`Handling "Symfony\Component\Mailer\Messenger\SendEmailMessage" failed: You must configure a "Symfony\Component\Mime\BodyRendererInterface" when a "Symfony\Bridge\Twig\Mime\NotificationEmail" instance has a text or HTML template set. (500 Internal Server Error)`

But when I try:

`$this->bodyRenderer->render($message);`

I get the error:

`TypeError: Symfony\Bridge\Twig\Mime\BodyRenderer::render(): Argument #1 ($message) must be of type Symfony\Component\Mime\Message, Symfony\Component\Notifier\Message\EmailMessage given`

### How to reproduce

NewNotificationByEmail.php

```
...
class NewNotificationByEmail implements NewNotificationInterface
{
    ...

    public function __invoke(array $pdfsName, int $entityId, string $entityToken, string $clientEmail, string $template) 
    {
        ...
        try {
            $clientNotification = new ClientNewNotificationByEmail($entityToken, $attachments, $template);
            $this->notifier->send($clientNotification, new Recipient($clientEmail));
        } catch (Exception $e) {
            $this->logger->critical('Failed to notify client', ['error'=>$e->getMessage(), 'code'=>$e->getCode(), 'file'=>$e->getFile(), 'line'=>$e->getLine(), 'id'=>$entityId]);
            throw $e;
        }
    }
}
```

ClientNewNotificationByEmail.php

```
...
class ClientNewNotificationByEmail extends Notification implements EmailNotificationInterface {
    ...

    public function asEmailMessage(EmailRecipientInterface $recipient, string $transport = null): ?EmailMessage
    {
        $message = EmailMessage::fromNotification($this, $recipient, $transport);

        $message->getMessage()
            ->htmlTemplate($this->template)
            ->context(['entityToken' => $this->entityToken, 'footer_text'=>null])
            ->importance('')
        ;

        foreach ($this->attachments as $att) {
            $message->getMessage()->attachFromPath($att['path'], $att['name'], 'application/pdf');
        }

        return $message;
    }
}
```

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #48472
| License       | MIT
| Doc PR        | n/a

Alternative to #48481

### CODE REVIEW

**Review by ThomasTr — APPROVED** [review]
Works for me

**Comment by xabbuh** [comment]
looks like we need to bump the minimum allowed version of `symfony/mime` in the Mailer component

File: src/Symfony/Component/Mailer/composer.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
         "psr/event-dispatcher": "^1",
         "psr/log": "^1|^2|^3",
         "symfony/event-dispatcher": "^5.4|^6.0",
-        "symfony/mime": "^6.2",
+        "symfony/mime": "^6.2.1",
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/composer.json]: no needed

File: src/Symfony/Component/Mailer/composer.json [hunk 2]
```diff
@@ -33,7 +33,7 @@
     "conflict": {
         "symfony/http-kernel": "<5.4",
         "symfony/messenger": "<6.2",
-        "symfony/mime": "<6.2"
+        "symfony/mime": "<6.2.1"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/composer.json]: the conflict should be with symfony/twig-bridge

File: src/Symfony/Bridge/Twig/Mime/NotificationEmail.php [hunk 3]
```diff
@@ -225,15 +236,17 @@ private function getExceptionAsString(\Throwable|FlattenException $exception): s
...
     /**
      * @internal
      */
     public function __unserialize(array $data): void
     {
-        if (3 === \count($data)) {
+        if (4 === \count($data)) {
+            [$this->context, $this->theme, $this->rendered, $parentData] = $data;
+        } elseif (3 === \count($data)) {
             [$this->context, $this->theme, $parentData] = $data;
```
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Mime/NotificationEmail.php:252]: Is worth adding a comment as done two lines below or should it be updated to cover this new case?

File: src/Symfony/Bridge/Twig/Mime/TemplatedEmail.php [hunk 4]
```diff
@@ -67,6 +67,18 @@ public function getContext(): array
         return $this->context;
     }
 
+    public function isRendered(): bool
+    {
+        return null === $this->htmlTemplate && null === $this->textTemplate;
```
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Mime/TemplatedEmail.php:72]: Why not using a boolean property as done in the other class?
Setting those to `null` in the next method could lead to inconsistency?

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Mailer/Transport/AbstractTransport.php [hunk 5]
```diff
@@ -76,7 +76,7 @@ public function send(RawMessage $message, Envelope $envelope = null): ?SentMessa
             $envelope = $event->getEnvelope();
             $message = $event->getMessage();
 
-            if ($message instanceof TemplatedEmail && ($message->getTextTemplate() || $message->getHtmlTemplate())) {
+            if ($message instanceof TemplatedEmail && !$message->isRendered()) {
```
Comment (HeahDude) [inline src/Symfony/Component/Mailer/Transport/AbstractTransport.php:79]: The behavior seems different here now, the error may be that the template has already been rendered.

**Pull request merged**